source/server/http/admin.cc

```diff
@@ -514,6 +514,51 @@ Http::Code AdminImpl::handlerCpuProfiler(absl::string_view url, Http::HeaderMap&
   return Http::Code::OK;
 }
 
+Http::Code AdminImpl::handlerHeapProfiler(absl::string_view url, Http::HeaderMap&,
+                                          Buffer::Instance& response, AdminStream&) {
+  if (!Profiler::Heap::profilerEnabled()) {
+    response.add("The current build does not support heap profiler");
+    return Http::Code::NotImplemented;
+  }
+
+  Http::Utility::QueryParams query_params = Http::Utility::parseQueryString(url);
+  if (query_params.size() != 1 || query_params.begin()->first != "enable" ||
+      (query_params.begin()->second != "y" && query_params.begin()->second != "n")) {
+    response.add("?enable=<y|n>\n");
+    return Http::Code::BadRequest;
+  }
+
+  Http::Code res = Http::Code::OK;
+  bool enable = query_params.begin()->second == "y";
+  if (enable) {
+    if (Profiler::Heap::isProfilerStarted()) {
+      response.add("Fail to start heap profiler: already started");
+      res = Http::Code::BadRequest;
+    } else if (!Profiler::Heap::startProfiler(profile_path_)) {
+      NOT_IMPLEMENTED_GCOVR_EXCL_LINE
```

Comment:
Here @zuercher

Reply:
This macro (NOT_IMPLEMENTED_GCOVR_EXCL_LINE) will *terminate the Envoy process*. If that's what you intend, remove the rest of the lines in this block, as they will never execute.

Reply:
I understand it. The reason it's safe is that startProfiler doesn't return false as long as the envoy build supports heap profiling. For envoy build without heap profiling, this branch never hit for other reason(See outer block)

The intention for this `NOT_IMPLEMENTED_GCOVR_EXCL_LINE` is saying, semnaticly it's a valid branch and we will implement in future PRs, but in current implementation it is deadcode.

I have the testcase to guarantee current behavior. 

What do you think? Is there any other approach to express this idea?

Reply:
The thing is you have implemented handling failure in startProfiler, even if it can't happen now. If someone comes along and modifies startProfiler's behavior without looking carefully at this class, they could create a situation where a running envoy can be terminated by mistake. So I still think you should remove the NOT_IMPLEMENTED_GCOVR_EXCL_LINE macro and let the (currently unreachable) error handling code stand with the comments indicating that it's not reach in code coverage tests.

Alternatively, you could also switch startProfiler to be a void method and remove the conditional (which is the way stopProfiler is called).

Reply:
You are right. If the new code randomly returns true at startProfiler, the unittest may not catch it and it will crash the envoy in production. 
I will follow your first approach.